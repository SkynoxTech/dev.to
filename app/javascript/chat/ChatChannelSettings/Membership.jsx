import { h } from 'preact';
import PropTypes from 'prop-types';
import { defaulMembershipPropType } from '../../common-prop-types/membership-prop-type';
import { Button } from '@crayons';

/**
 *
 * This component render the membership section
 *
 * @param {object} props
 * @param {object} membership
 * @param {function} removeMembership
 * @param {function} membershipType
 * @param {function} chatChannelAcceptMembership
 * @param {string} currentMembershipRole
 *
 * @compoent
 *
 * @example
 */
export default function Membership({
  membership,
  removeMembership,
  membershipType,
  chatChannelAcceptMembership,
  currentMembershipRole,
}) {
  return (
    <div className="flex items-center">
      <a
        href={`/${membership.username}`}
        className="chatmessagebody__username--link"
        target="_blank"
        rel="noopener noreferrer"
        data-content="sidecar-user"
      >
        <span className="crayons-avatar crayons-avatar--l mr-3">
          <img
            className="crayons-avatar__image align-middle"
            src={membership.image}
            alt={`${membership.name} profile`}
          />
        </span>
        <span className="mr-2 user_name">{membership.name}</span>
      </a>
      {membershipType === 'requested' ? (
        <Button
          className="crayons-btn crayons-btn--icon-rounded crayons-btn--ghost add-membership"
          type="button"
          onClick={chatChannelAcceptMembership}
          data-membership-id={membership.membership_id}
        >
          +
        </Button>
      ) : null}
      {membership.role !== 'mod' && currentMembershipRole === 'mod' ? (
        <Button
          className="crayons-btn crayons-btn--icon-rounded crayons-btn--ghost remove-membership"
          type="button"
          onClick={removeMembership}
          data-membership-id={membership.membership_id}
          data-membership-status={membership.status}
        >
          x
        </Button>
      ) : null}
    </div>
  );
}

Membership.propTypes = {
  membership: PropTypes.objectOf(defaulMembershipPropType).isRequired,
  removeMembership: PropTypes.func.isRequired,
  membershipType: PropTypes.func.isRequired,
  chatChannelAcceptMembership: PropTypes.func.isRequired,
  currentMembershipRole: PropTypes.string.isRequired,
};
