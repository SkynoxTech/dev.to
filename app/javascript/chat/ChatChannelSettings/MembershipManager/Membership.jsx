import { h } from 'preact';
import PropTypes from 'prop-types';

import ConfigImage from '../../../../assets/images/three-dots.svg';
import adminEmoji from '../../../../assets/images/emoji/apple-fire.png';

const Membership = ({
  membership,
  currentMembership,
  removeMembership,
  handleUpdateMembershipRole,
}) => {
  const addAsModButton =
    membership.role === 'member' ? (
      <button
        className="crayons-btn crayons-btn--ghost remove-membership p-2"
        type="button"
        onClick={handleUpdateMembershipRole}
        data-membership-id={membership.membership_id}
        data-role="mod"
      >
        Add as Mod
        <span>
          <img
            src={adminEmoji}
            alt="admin emoji"
            data-content="admin emoji"
            className="admin-emoji-button mx-2"
            title="MOD"
          />
        </span>
      </button>
    ) : null;

  const addAsMemberButton =
    membership.role === 'mod' ? (
      <button
        className="crayons-btn crayons-btn--ghost remove-membership p-2"
        type="button"
        onClick={handleUpdateMembershipRole}
        data-membership-id={membership.membership_id}
        data-role="member"
      >
        Add as Member
      </button>
    ) : null;

  const removeMembershipButton =
    membership.role === 'member' ? (
      <button
        className="crayons-btn crayons-btn--ghost  crayons-btn--ghost-danger remove-membership p-2"
        type="button"
        onClick={removeMembership}
        data-membership-id={membership.membership_id}
        data-membership-status={membership.status}
      >
        Remove
      </button>
    ) : null;

  const dropdown =
    currentMembership.role === 'mod' ? (
      <div className="membership-actions">
        <span className="membership-management__dropdown-button">
          <img
            src={ConfigImage}
            alt="channel config"
            data-content="drop-down-image"
            className="w-25"
          />
        </span>

        <div className="membership-management__dropdown-memu">
          {addAsModButton}
          {addAsMemberButton}
          {removeMembershipButton}
        </div>
      </div>
    ) : null;

  return (
    <div className="flex items-center my-3 member-list-item justify-content-between">
      <div className="w-100">
        <a
          href={`/${membership.username}`}
          className="chatmessagebody__username--link"
          target="_blank"
          rel="noopener noreferrer"
          data-content="sidecar-user"
        >
          <span className="crayons-avatar crayons-avatar--l mr-3">
            <img
              className="crayons-avatar__image align-middle"
              role="presentation"
              src={membership.image}
              alt={`${membership.name} profile`}
            />
          </span>
          <span className="mr-2 user_name">{membership.name}</span>
          <span>
            {membership.role === 'mod' ? (
              <img
                src={adminEmoji}
                alt="admin emoji"
                data-content="admin emoji"
                className="admin-emoji"
                title="MOD"
              />
            ) : null}
          </span>
        </a>
      </div>
      <div className="flex-shrink-1">{dropdown}</div>
    </div>
  );
};

Membership.propTypes = {
  membership: PropTypes.objectOf(
    PropTypes.shape({
      name: PropTypes.string.isRequired,
      membership_id: PropTypes.number.isRequired,
      user_id: PropTypes.number.isRequired,
      role: PropTypes.string.isRequired,
      image: PropTypes.string.isRequired,
      username: PropTypes.string.isRequired,
      status: PropTypes.string.isRequired,
    }),
  ).isRequired,
  currentMembership: PropTypes.isRequired,
  removeMembership: PropTypes.func.isRequired,
  handleUpdateMembershipRole: PropTypes.func.isRequired,
};

export default Membership;
