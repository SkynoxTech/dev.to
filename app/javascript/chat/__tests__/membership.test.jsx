import { h } from 'preact';
import render from 'preact-render-to-json';
import { shallow } from 'preact-render-spy';
import Membership from '../ChatChannelSettings/Membership';

const data = {
  membership: {
    name: 'test user',
    username: 'testusername',
    user_id: '1',
    membership_id: '2',
    role: 'mod',
    status: 'active',
    image: '',
  },
  membershipType: 'active',
  currentMembershipRole: 'mod',
};

const getMembership = (membershipData) => (
  <Membership
    membership={membershipData.membership}
    membershipType={membershipData.membershipType}
    currentMembershipRole={membershipData.currentMembershipRole}
  />
);

describe('<Membership />', () => {
  it('should render and test snapshot', () => {
    const tree = render(getMembership(data));
    expect(tree).toMatchSnapshot();
  });

  it('should have the elements', () => {
    const context = shallow(getMembership(data));

    expect(context.find('.user_name').text()).toEqual(data.membership.name);
  });
});
