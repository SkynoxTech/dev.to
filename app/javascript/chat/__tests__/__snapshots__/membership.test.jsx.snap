// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Membership /> should render and test snapshot 1`] = `
<div
  class="flex items-center"
>
  <a
    class="chatmessagebody__username--link"
    data-content="sidecar-user"
    href="/testusername"
    rel="noopener noreferrer"
    target="_blank"
  >
    <span
      class="crayons-avatar crayons-avatar--l mr-3"
    >
      <img
        alt="test user profile"
        class="crayons-avatar__image align-middle"
        role="presentation"
        src=""
      />
    </span>
    <span
      class="mr-2 user_name"
    >
      test user
    </span>
  </a>
</div>
`;
